### ISSUE TITLE
Add support of ED25519-SHA256 for signing DKIM Keys

### ISSUE THREAD

**Issue by saldru** [mailcow/mailcow-dockerized#2972]
Hi everyone,

I ask for a feature that could be interesting. Actually the mailCow suite only support DKIM keys signed with RSA through the web UI. Since [04/02/2019](https://github.com/rspamd/rspamd/pull/2739), Rspamd added support for signing DKIM keys with ED25519 with SHA256 regarding the new [RFC 8463](https://tools.ietf.org/html/rfc8463). 

So it could be good to add the option of generating/importing and supporting DKIM keys signed with ED25519-SHA256 in addition of RSA keys.

Thank you for you time :)
Saldru

**Comment by andryyy** [comment]
I will add it as FR, but I don't think we will add it very soon. Not many servers can actually validate them, btw.

We could double-sign, but that's just overly complicated for most people running a mailcow. Same with this multi-cert stuff with RSA and ECDSA. Multiple different TLSA records per domains sound fun for most people, who chose mailcow to reduce the hassle.
We will eventually move forward, yes, but I don't think we should add more hacky solutions for something not yet fully supported or established.

I think DKIM with ED25519 is not too far away though. :)

Thanks for your FR!

**Comment by saldru** [comment]
Hi @andryyy, 
thank you for your reply ! 

Yeah I know that it's new and it's better for mails when the policy is the most open and not too "modern" to be able to receive and send mail from/to everyone. :)

But I think it's a feature that can say to everyone that "mailcow support it !" and yeah I know it's not in the top of priority. :)

**Comment by ghost** [comment]
> I think DKIM with ED25519 is not too far away though. :)

@andryyy Is support for signing mails with ED25519 DKIM keys still on the roadmap for mailcow?

**Comment by andryyy** [comment]
Yes, in the future. We will do dual signing at some point.

As of today it is pretty much not checked at all...

**Comment by andryyy** [comment]
It is.

**Comment by JJ1LFC** [comment]
Hi, It's been 4yrs now. We should not keep messing up our TXT RR with very long RSA 1024/2048 bit pubkey - or should we go to RSA 4096 seriously? It's a stupid idea. Let's boost the entire world to use shorter elliptic curve algo.

**Comment by martinjgrunwald** [comment]
+1

**Comment by wgentine** [comment]
+1

**Comment by JensHumke** [comment]
ED25519 + RSA ? If we wait any longer we get PQC DKIM support

**Comment by apio-sys** [comment]
I don't think this is a good idea as yet. Adoption of it is very poor (doesn't work for GMail, O365, Yahoo or iCloud according to this: https://redsift.com/blog/ed25519-dkim-support-weak-keys). Even though we should not not do it because the giants don't support it, but in this case would give complex setups which will make mailcow more complex to manage. The "only" save way today is going for RSA2048, even beyond that with RSA4096 you can face compatibility issues (DNS servers not excepting such long text entries in their GUI, "legacy" mail servers (incl. O365) not accepting those keys, ...).

IMHO, it's more urgent to work on getting rid of RSA in our main certs for Postfix/Dovecot/NGinx. I have posted a PR for that here #7250 a while back. It's a simple change with immediate gain and very low risk.

**Comment by JensHumke** [comment]
RSA4096 is no problem at least what I can tell. Ed25519 sure has some adoption issues but we have to move to PQC anyways so we should at least offer it as an option.

**Comment by apio-sys** [comment]
As optional why not. RSA4096 can definitely be a problem for the reasons I mentioned above. Anyway, you can already use that through the admin GUI but you risk delivery issue with those "giants"...

**Comment by JensHumke** [comment]
Yeah, it was already difficult to get my E-mails accepted in the first place (hosting at home). Well I wonder if there will be a shift maybe 2027.

**Comment by JensHumke** [comment]
Hi again,

I would like to add a more concrete follow-up after testing this on my own mailcow setup.

My requirement is not Ed25519-only DKIM. I fully understand the compatibility concerns with large providers such as Gmail, Microsoft 365, Yahoo/iCloud and others.

What I would need is optional **dual DKIM signing**:

- RSA-XXXX as the mandatory compatibility signature
- Ed25519 as an additional modern signature

So outgoing mail would ideally contain both signatures:

```txt
DKIM-Signature: a=rsa-sha256; d=example.com; s=dkim; ...
DKIM-Signature: a=ed25519-sha256; d=example.com; s=ed25519; ...

I tested a few approaches with the current mailcow/Rspamd setup.

The standard mailcow DKIM path works reliably with RSA through the Redis-based key management:

use_redis = true
key_prefix = DKIM_PRIV_KEYS
selector_prefix = DKIM_SELECTORS

However, trying to add an additional Ed25519 selector outside the mailcow UI does not seem to be update-safe or production-safe.

What I observed:

publishing an additional ed25519._domainkey DNS record alone is not useful unless outgoing mail is actually signed with s=ed25519
trying to add custom Rspamd local.d DKIM selector configuration resulted in unexpected/fragile behaviour
disabling/bypassing the mailcow Redis-based DKIM path through override.d caused DKIM signing to stop in my test, so this is not something I would consider safe for production
mailcow currently appears to manage one active DKIM selector/key per domain through the UI/key management model

My questions are:

Is RSA + Ed25519 dual DKIM signing currently possible in mailcow in an update-safe way?
If not, what is the main blocker?
mailcow UI/key model?
Redis DKIM key storage?
Rspamd selector handling?
ARC/DKIM interaction?
receiver compatibility concerns?
Would the project consider an optional advanced mode for dual signing?

A safe implementation could be:

keep RSA mandatory when Ed25519 is enabled
allow an additional Ed25519 selector/key for the same domain
show both DNS records in the UI
sign outbound mail with both selectors
keep the existing RSA/ARC behaviour as the compatibility baseline
clearly mark Ed25519 as optional/advanced due to receiver compatibility

I am not asking for Ed25519-only DKIM to become the default. I agree that this would be risky today. But optional dual signing would be useful for operators who want to adopt newer standards early while keeping RSA compatibility.

If this is currently not feasible, an official explanation would also be helpful, so administrators know whether this is a mailcow limitation, an Rspamd integration limitation, or an intentional compatibility decision.

Thanks again for your work on mailcow.
